### ISSUE TITLE
upload: progress indication for "digesting"?

### ISSUE THREAD

**Issue by yarikoptic** [dandi/dandi-cli#400]
"digesting" could take comparable to upload time, in particular in the case of the parallel `upload`s. Although with caching we at least would not bother redoing it twice if did recently, it would be nice if we have a way to report back to the user on the progress of digesting.  For upload we have a dedicated column `upload` which shows `%`.  But I wonder if we better have a dedicated column "progress" which would be populated (and either cleared or report time on how long it took after it is done) used by whatever 'status' column reports (digesting, or uploading, etc).

**Comment by yarikoptic** [comment]
I will leave UX decision to you (a separate column or just reuse existing)

**Comment by jwodder** [comment]
@yarikoptic I believe `joblib.Memory` (on which fscacher depends) uses pickle to store data, and iterators aren't pickleable.  In order to implement this, we'd need a way to query the cache to see whether it already has an entry for a given path and a way to manually insert entries in the cache, neither of which seem to be supported by `joblib`.

**Comment by satra** [comment]
perhaps use pydra instead  and store the cache directory locally somewhere.

**Comment by jwodder** [comment]
@satra Exactly what relevant features does pydra have?  Digest progress is the easy part; caching it on the filesystem is the hard part.

**Comment by satra** [comment]
that's exactly what pydra would provide in addition to parallelization of a process :) 

i would suggest going through the tutorial: https://github.com/nipype/pydra-tutorial

**Comment by jwodder** [comment]
@satra I see that that provides caching of functions, but I don't see how one would get digest progress information out of it.  Keep in mind that we're currently digesting using sha256, which isn't parallelizable, unlike the contemplated Merkel tree hash.

**Comment by yarikoptic** [comment]
@satra FWIW pydra is relatively heavy of a dependency (and dependencies down) to use just for caching.  Even joblib is somewhat too much since we use just memoization part... Also I see it depends on cloudpickle -- is that is what it uses for pickling? its [README](https://github.com/cloudpipe/cloudpickle) warns 

```
Cloudpickle can only be used to send objects between the exact same version of Python.

Using cloudpickle for long-term object storage is not supported and strongly discouraged.
```
which possibly makes it suboptimal (although I am not sure how good we are now for sure anyways: https://github.com/con/fscacher/issues/36)

**Comment by yarikoptic** [comment]
@jwodder , with all the generators it needs more thinking on how to wrap it all up ... meanwhile - having support for them in fscacher would be useful, so filed https://github.com/con/fscacher/issues/37

**Comment by satra** [comment]
perhaps i misunderstood what you meant by

> caching it on the filesystem is the hard part.

are you saying caching the progressing state so you could start at whatever point it stopped? or are you caching the entire hash?  indeed pydra is not suited for progressive filesystem caching. but whenever you run a function, if it's completed the local cache is like a memoization on disk.

**Comment by jwodder** [comment]
@satra We need the cache to be a datastore mapping tuples to bytes that (a) can be freely queried & modified, not just used to dumbly memoize a function, and (b) is threading- and multiprocessing-safe.  I originally envisioned something like how joblib.Memory works, where pickled files are laid out in a directory tree based on hashes of the inputs or something like that, which would mean coming up with our own layout scheme, though now that I think about it more, we might just be able to get away with a sqlite database....

**Comment by satra** [comment]
i think i'm still not fully following what the cache stores i.e what the contents of tuples are what bytes, but sounds like you have figured out a solution :)

**Comment by jwodder** [comment]
@yarikoptic Regarding the pyout columns: I've renamed "upload" to "pct" and used it for both digest progress and upload progress, but it appears that the "size" and "pct" columns both stop updating after digesting is done, even though they should start counting from zero again while uploading.  Do you know why this is?

**Comment by yarikoptic** [comment]
no problem -- `summary` could be customized and we could exclude progress reporting. ATM it is
```python
        "message": dict(
            color=dict(
                re_lookup=[["^exists", "yellow"], ["^(failed|error|ERROR)", "red"]]
            ),
            aggregate=counts,
        ),
```
so you could just provide a `counts_no_progress` (or just `lambda` or `filter` right there on top of counts) which would first filter entries

**Comment by jwodder** [comment]
@yarikoptic I tried [this](https://github.com/dandi/dandi-cli/pull/465/commits/73aacb71e491fbfb6dec1539297349b692c675e2), but it didn't make a difference.

**Comment by yarikoptic** [comment]
re the overall approach... I think it could also be done quite non-intrusively (although not sure if having a callback bound to Digester would have some negative side effect on fscacher) via use of the `generator_from_callback` we have and used for reporting progress from `upload` for girder backend: 

add optional callback to Digester, wrap call to a digester into `generator_from_callback` and iterate it while getting the final result from `StopIteration` exception since that is where the final result is provided by the `generator_from_callback`
```python
                ret = func(callback)
                raise StopIteration(ret) if ret is not None else StopIteration
```

**Comment by yarikoptic** [comment]
> @yarikoptic I tried [this](https://github.com/dandi/dandi-cli/pull/465/commits/73aacb71e491fbfb6dec1539297349b692c675e2), but it didn't make a difference.

didn't try to debug but may be because `%` is [yielded in "status"](https://github.com/dandi/dandi-cli/pull/465/files#diff-b338a27af45ae7897ca46a8130b866561898827c8cbdac88985370d6827a909cR148) and not "message" field [which you are filtering?](https://github.com/dandi/dandi-cli/pull/465/files#diff-28d9bc8a6606b4a5ea8161e32c1aa050edaf50bf81d22d5aed69917127e4ea60R176)

please also see https://github.com/dandi/dandi-cli/issues/400#issuecomment-797601538 on may be a simpler path toward needed functionality and avoiding adding a new dependency (diskcache) with a known NFS issue

**Comment by jwodder** [comment]
@yarikoptic That was it.  The summary is still rather flickery, though.

**Comment by jwodder** [comment]
@yarikoptic If I'm understanding `generator_from_callback()` correctly, you use it by creating a function `func` that takes a callback, and then you do `generator_from_callback(func)` to get an iterable of the return values from the callback.  Is that correct?  What are you envisioning as the `func` in this case?  It can't be `get_digest()`, because then the callback argument would mess with argument caching.  It can't be a function called by `get_digest()`, because then the generator would be inside `get_digest()`, and we're back to square one with the impossibility of returning an iterator from a cached function.  It can't be a function that calls `get_digest()`, because `get_digest()` doesn't return progress information and can't without being uncacheable.

**Comment by yarikoptic** [comment]
> It can't be `get_digest()`, because then the callback argument would mess with argument caching

we can parametrize `PersistentCache` with specific `exclude_kwargs=None|iterable` (e.g. in our case `exclude_kwargs=['callback']`) to be excluded from the signature used for caching . Then callback could be passed through without affecting caching

**Comment by jwodder** [comment]
@yarikoptic The good news is that joblib's Memory.cache has an `ignore` parameter that can be used to implement this.  The bad news is that the implementation does not work with fscacher's repeated function-wrapping.  [I filed a bug report.](https://github.com/joblib/joblib/issues/1164)

**Comment by yarikoptic** [comment]
FTR: PR was merged and we are waiting for joblib release, asked: https://github.com/joblib/joblib/pull/1165#issuecomment-832021219

**Comment by kabilar** [comment]
Our LINC collaborator @dstansby is uploading a large Zarr with the LINC Client and has suggested that we provide feedback for the uploads.  Below I have copied the text from the original issue (https://github.com/lincbrain/linc-cli/issues/45).

> I'm currently uploading a large (300 GB) dataset, or at least I think I am... there's a Python process using my outgoing internet at 200Mbps, but all that the lincbrain cli has told me is:
> ```
> $ lincbrain upload
> 2024-05-02 15:20:14,885 [    INFO] Found 6 files to consider
> ```
> 
> It would be nice if there was some sort of progress bar, or at a minimum a message saying the CLI was starting a file upload.

Thank you.

cc @aaronkanzer

**Comment by yarikoptic** [comment]
I guess we would need to get back to postponed
- https://github.com/dandi/dandi-cli/pull/465
but meanwhile, @jwodder do you think we could at least have overall status "digesting" to be displayed without progress for a file (zarr or not) before we start digesting it for upload?

**Comment by kabilar** [comment]
Thanks @jwodder.  To clarify, does `dandi upload` also have this option (`--log-level debug`)?  I didn't see it in the [docs](https://dandi.readthedocs.io/en/latest/cmdline/upload.html).

**Comment by jwodder** [comment]
@kabilar It's an option of the top-level [`dandi` command](https://dandi.readthedocs.io/en/latest/cmdline/dandi.html), not a subcommand option.

**Comment by kabilar** [comment]
Perfect, thank you.  Will test out today.
